### PR TITLE
update_roots on pool to keep it consistent after reconciliation

### PR DESCRIPTION
**Pull request by antiochp**
This fixes the issue I was seeing in #181 
We can now successfully spend a zero confirmation change output and it will get mined after a couple of blocks.
We currently only mine `roots` from the pool - this PR ensures `roots` is consistent after we call `reconcile_block`.

* introduced `update_roots` to ensure the pool is consistent after reconciliation
* remove duplicates before marking txs (leverage sets here)
* use HashSet instead of HashMap when marking txs

### CODE REVIEW

File: pool/src/graph.rs [hunk 1]
```diff
@@ -179,6 +180,31 @@ impl DirectedGraph {
 		}
 	}
 
+	/// Promote any non-root vertices to roots based on current edges.
+	/// For a given tx, if there are no edges with that tx as destination then it is a root.
+	pub fn update_roots(&mut self) {
+		let mut new_roots = vec![];
+		for x in &self.vertices {
+			let edges = &self.edges.values().collect::<Vec<_>>();
```
Comment (MoaningMyrtle) [inline pool/src/graph.rs]: I think you can move this outside the for loop to avoid allocating for every vertex.

File: pool/src/graph.rs [hunk 2]
```diff
@@ -179,6 +180,31 @@ impl DirectedGraph {
...
+				Some(_) => (),
+				None => new_roots.push(x.clone()),
+			}
+		}
+
+		for x in new_roots {
+			self.roots.push(x.clone());
+		}
+
+		let mut new_vertices = vec![];
```
Comment (MoaningMyrtle) [inline pool/src/graph.rs]: I think you can avoid allocating a new vertices vector here and also avoid comparing against all of `roots` by reusing `new_roots`. Something like (not tested):

```
		for x in new_roots {
			self.roots.push(x.clone());
                        match self.vertices.iter().position(|v| v.transaction_hash == x.transaction_hash) {
                                   Some(i) => self.vertices.swap_remove(i),
                                    None=> (),
                         }
                  }
```

Even better, since you know where vertices are matching from the above loop (L187) you could do without this second position search altogether- although the ways I can think of to do that now are messy (would be really nice with Rust nightly's `drain_filter`, which isn't stable yet). That feels like an optimization that can wait, though.

**Comment by antiochp** [comment]
Thanks for the feedback - good suggestions.
This was definitely more "make it work" and less "make it fast".
But let me take another pass tomorrow and I'll clean some of this up. I think adding the HashMaps so we can index the sources and destinations makes a lot of sense.

**Comment by antiochp** [comment]
@MoaningMyrtle @ignopeverell - this is ready for review.
A *lot* less iterating and allocating going on in `update_roots` now.
I also noticed during testing that `conflicting_txs` and `conflicting_outputs` contained a lot of duplicate hashes (multiple outputs of a given tx etc.) So made some changes to de-dupe these during the "mark" process.

**Comment by antiochp** [comment]
Not sure why TravisCI failed here - it failed on "address in use" during the `grin_grin` tests.
@ignopeverell can you restart the tests in TravisCI? (I do not have permission).
